### PR TITLE
ftr (billable metrics): unique count aggregation refactoring

### PR DESCRIPTION
**Pull request by lovrocolic**
## Context

As part of new charge pay in advance feature, unique count aggregation is refactored and computation of pay in advance event is changed

## Description

This PR covers few parts:

- aggregation for unique count BM has changed since currently we support operation types
- computation of pay in advance event has changed since we need to take into account maximum in current period in order to calculate fair price for the user

### CODE REVIEW

File: app/services/billable_metrics/aggregations/unique_count_service.rb [hunk 1]
```diff
@@ -4,29 +4,41 @@ module BillableMetrics
...
+      def compute_pay_in_advance_aggregation
         return 0 unless event
         return 0 if event.properties.blank?
 
-        existing_property = events
-          .where("#{sanitized_field_name} = '#{event.properties[billable_metric.field_name]}'")
-          .where.not(id: event.id)
-          .any?
+        unless previous_event
+          res = operation_type == :add ? 1 : 0
```
Comment (jdenquin) [inline app/services/billable_metrics/aggregations/unique_count_service.rb:22]: Style/TernaryParentheses: Use parentheses for ternary expressions with complex conditions.

File: app/services/billable_metrics/aggregations/unique_count_service.rb [hunk 2]
```diff
@@ -4,29 +4,41 @@ module BillableMetrics
...
-        # NOTE: bill only property that have not been received in the period yet
-        return 0 if existing_property
+          return res
+        end
 
-        1
+        old_aggregation = BigDecimal(previous_event.metadata['current_aggregation'])
+        old_max = BigDecimal(previous_event.metadata['max_aggregation'])
+
+        current_aggregation = operation_type == :add ? (old_aggregation + 1) : (old_aggregation - 1)
```
Comment (jdenquin) [inline app/services/billable_metrics/aggregations/unique_count_service.rb:31]: Style/TernaryParentheses: Use parentheses for ternary expressions with complex conditions.

File: app/services/billable_metrics/aggregations/unique_count_service.rb [hunk 3]
```diff
@@ -4,29 +4,41 @@ module BillableMetrics
...
+
+        if current_aggregation > old_max
+         handle_event_metadata(current_aggregation:, max_aggregation: current_aggregation)
```
Comment (jdenquin) [inline app/services/billable_metrics/aggregations/unique_count_service.rb:34]: Layout/IndentationWidth: Use 2 (not 1) spaces for indentation. (https://rubystyle.guide#spaces-indentation)

File: app/services/billable_metrics/aggregations/unique_count_service.rb [hunk 4]
```diff
@@ -4,29 +4,41 @@ module BillableMetrics
...
+
+         1
+        else
+         handle_event_metadata(current_aggregation:, max_aggregation: old_max)
```
Comment (jdenquin) [inline app/services/billable_metrics/aggregations/unique_count_service.rb:38]: Layout/IndentationWidth: Use 2 (not 1) spaces for indentation. (https://rubystyle.guide#spaces-indentation)

File: app/services/billable_metrics/aggregations/unique_count_service.rb [hunk 5]
```diff
@@ -39,6 +51,96 @@ def running_total(options)
...
+            .where('quantified_events.added_at::timestamp(0) >= ?', from_datetime)
+            .where('quantified_events.added_at::timestamp(0) <= ?', to_datetime)
+            .order(created_at: :desc)
+
+          query
+            .where('quantified_events.removed_at::timestamp(0) IS NULL')
+            .or(
+              query
+                .where('quantified_events.removed_at::timestamp(0) >= ?', from_datetime)
+                .where('quantified_events.removed_at::timestamp(0) <= ?', to_datetime)
```
Comment (jdenquin) [inline app/services/billable_metrics/aggregations/unique_count_service.rb:77]: Style/TrailingCommaInArguments: Put a comma after the last parameter of a multiline method call. (https://rubystyle.guide#no-trailing-params-comma)

File: app/services/billable_metrics/aggregations/unique_count_service.rb [hunk 6]
```diff
@@ -39,6 +51,96 @@ def running_total(options)
...
+
+      def compute_aggregation
+        ActiveRecord::Base.connection.execute(aggregation_query).first['aggregation_result']
+      end
+
+      def aggregation_query
+        queries = [
+          # NOTE: Billed on the full period. We will replace 1::numeric with proration_coefficient::numeric
+          # in the next part
+          persisted.select("SUM(1::numeric)").to_sql,
```
Comment (jdenquin) [inline app/services/billable_metrics/aggregations/unique_count_service.rb:101]: Style/StringLiterals: Prefer single-quoted strings when you don't need string interpolation or special symbols. (https://rubystyle.guide#consistent-string-literals)

File: app/services/billable_metrics/aggregations/unique_count_service.rb [hunk 7]
```diff
@@ -39,6 +51,96 @@ def running_total(options)
...
+
+          # NOTE: Added during the period, We will replace 1::numeric with proration_coefficient::numeric
+          # in the next part
+          added.select("SUM(1::numeric)").to_sql,
```
Comment (jdenquin) [inline app/services/billable_metrics/aggregations/unique_count_service.rb:105]: Style/StringLiterals: Prefer single-quoted strings when you don't need string interpolation or special symbols. (https://rubystyle.guide#consistent-string-literals)

File: app/services/quantified_events/create_or_update_service.rb [hunk 8]
```diff
@@ -30,7 +30,8 @@ def matching_billable_metric?
     delegate :customer, :subscription, :organization, to: :event
 
     def event_operation_type
-      event.properties['operation_type']&.to_sym
+      operation_type = event.properties['operation_type']&.to_sym
+      operation_type.nil? && matching_billable_metric&.unique_count_agg? ? :add : operation_type
```
Comment (jdenquin) [inline app/services/quantified_events/create_or_update_service.rb:34]: Style/TernaryParentheses: Use parentheses for ternary expressions with complex conditions.

File: app/services/quantified_events/validate_creation_service.rb [hunk 9]
```diff
@@ -26,7 +26,10 @@ def valid?
     delegate :customer, to: :subscription
 
     def operation_type
-      @operation_type ||= args.dig('properties', 'operation_type')&.to_sym
+      @operation_type ||= begin
+        type = args.dig('properties', 'operation_type')&.to_sym
+        type.nil? && billable_metric.unique_count_agg? ? :add : type
```
Comment (jdenquin) [inline app/services/quantified_events/validate_creation_service.rb:31]: Style/TernaryParentheses: Use parentheses for ternary expressions with complex conditions.

File: spec/services/billable_metrics/aggregations/unique_count_service_spec.rb [hunk 10]
```diff
@@ -22,157 +33,319 @@
...
+          event: pay_in_advance_event,
+          external_subscription_id: subscription.external_id,
+          billable_metric:,
+        )
+      end
+
+      before { new_quantified_event }
 
-      it 'assigns zero as pay_in_advance aggregation' do
+      let(:properties) { { unique_id: '111' } }
```
Comment (jdenquin) [inline spec/services/billable_metrics/aggregations/unique_count_service_spec.rb:255]: RSpec/ScatteredLet: Group all let/let! blocks in the example group together. (https://www.rubydoc.info/gems/rubocop-rspec/RuboCop/Cop/RSpec/ScatteredLet)

File: spec/services/quantified_events/create_or_update_service_spec.rb [hunk 11]
```diff
@@ -80,6 +80,32 @@
       end
     end
 
+    context 'context without operation type for unique_count_agg' do
```
Comment (jdenquin) [inline spec/services/quantified_events/create_or_update_service_spec.rb:83]: RSpec/ContextWording: Context description should match /^when\b/, /^with\b/, or /^without\b/. (https://rspec.rubystyle.guide/#context-descriptions, https://www.rubydoc.info/gems/rubocop-rspec/RuboCop/Cop/RSpec/ContextWording)

File: app/services/billable_metrics/aggregations/unique_count_service.rb [hunk 12]
```diff
@@ -39,6 +51,96 @@ def running_total(options)
...
+
+          # NOTE: Added during the period, We will replace 1::numeric with proration_coefficient::numeric
+          # in the next part
+          added.select('SUM(1::numeric)').to_sql,
+        ]
+
+        "SELECT (#{queries.map { |q| "COALESCE((#{q}), 0)" }.join(' + ')}) AS aggregation_result"
+      end
+
+      def persisted
```
Comment (vincent-pochet) [inline app/services/billable_metrics/aggregations/unique_count_service.rb:111]:
```suggestion
      def persisted_query
```

File: app/services/billable_metrics/aggregations/unique_count_service.rb [hunk 13]
```diff
@@ -39,6 +51,96 @@ def running_total(options)
...
+        return QuantifiedEvent.none unless billable_metric.recurring?
+
+        base_scope
+          .where('quantified_events.added_at::timestamp(0) < ?', from_datetime)
+          .where('quantified_events.removed_at IS NULL OR quantified_events.removed_at::timestamp(0) > ?', to_datetime)
+      end
+
+      def added
```
Comment (vincent-pochet) [inline app/services/billable_metrics/aggregations/unique_count_service.rb:119]:
```suggestion
      def added_query
```

File: spec/services/billable_metrics/aggregations/unique_count_service_spec.rb [hunk 14]
```diff
@@ -22,157 +33,318 @@
...
+        )
+      end
 
-    it 'counts as zero' do
-      result = count_service.aggregate(from_datetime:, to_datetime:)
+      before { new_quantified_event }
 
-      expect(result.aggregation).to eq(0)
-      expect(result.count).to eq(0)
+      it 'returns the only the number of events ingested in current period' do
```
Comment (vincent-pochet) [inline spec/services/billable_metrics/aggregations/unique_count_service_spec.rb:104]:
```suggestion
      it 'returns only the number of events ingested in the current period' do
```

**Review by vincent-pochet — COMMENTED** [review]
Looks like a really good start :+1:

File: app/services/invoices/calculate_fees_service.rb [hunk 15]
```diff
@@ -105,7 +105,15 @@ def create_subscription_fee(subscription, boundaries)
...
-      subscription.plan.charges.where(pay_in_advance: false, invoiceable: true).each do |charge|
+      subscription
+        .plan
+        .charges
+        .joins(:billable_metric)
+        .where(invoiceable: true)
+        .where
+        .not(pay_in_advance: true, billable_metric: { recurring: false })
+        .each do |charge|
+
```
Comment (jdenquin) [inline app/services/invoices/calculate_fees_service.rb:116]: Layout/EmptyLinesAroundBlockBody: Extra empty line detected at block body beginning. (https://rubystyle.guide#empty-lines-around-bodies)

File: app/services/invoices/calculate_fees_service.rb [hunk 16]
```diff
@@ -105,7 +105,9 @@ def create_subscription_fee(subscription, boundaries)
     end
 
     def create_charges_fees(subscription, boundaries)
-      subscription.plan.charges.where(pay_in_advance: false, invoiceable: true).each do |charge|
+      subscription.plan.charges.includes(:billable_metric).where(invoiceable: true).each do |charge|
+        next if charge.pay_in_advance? && !charge.billable_metric.recurring?
```
[the remark this answers is not in the document]
Comment (lovrocolic) [reply app/services/invoices/calculate_fees_service.rb]: Done!

**Review by vincent-pochet — APPROVED** [review]

**Pull request merged**